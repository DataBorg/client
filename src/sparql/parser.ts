import type { SparqlQuery } from 'sparqljs';
import { Generator, Parser } from 'sparqljs';

export const sparqlParser = new Parser();
export const sparqlGenerator = new Generator();

export const sparqlQueryToString = (query: SparqlQuery): string => {
  const queryString = sparqlGenerator.stringify(query);
  return queryString;
};
