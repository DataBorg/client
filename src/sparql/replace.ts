import type {
  BgpPattern,
  BindPattern,
  ConstructQuery,
  Expression,
  GroupPattern,
  InsertDeleteOperation,
  OperationExpression,
  Pattern,
  Quads,
  Query,
  SparqlQuery,
  Term,
  Triple,
  Update,
} from 'sparqljs';
import type { QueryVariables } from 'types/client';

const isValueURI = (value: string | Term) => {
  if (typeof value === 'string') {
    return false;
  }

  return value.termType === 'NamedNode';
};

const handleVariableType = (value: string | Term) => {
  if (typeof value !== 'string' && value.termType !== undefined) {
    return value.termType;
  }

  return isValueURI(value) ? 'NamedNode' : 'Literal';
};

const handleVariableValue = (value: string | Term) => {
  if (typeof value === 'string') {
    return value;
  }

  return value.value;
};

const replaceTerm = ({
  term,
  name,
  value,
}: {
  term: { termType?: string; value?: string };
  name: string;
  value: any;
}) => {
  // ignore undefined values
  if (!term) {
    return;
  }
  // console.log('replace', { term, name, value });
  if (term.termType === 'Variable' && term.value === name) {
    term.termType = handleVariableType(value);
    term.value = handleVariableValue(value);
  } else if (term.termType === 'NamedNode' && term.value === `var://${name}`) {
    term.termType = handleVariableType(value);
    term.value = handleVariableValue(value);
  }
};

const replaceExpression = ({
  expression,
  variables,
}: {
  expression: Expression;
  variables: QueryVariables;
}): Expression => {
  // console.log(expression);
  if (Array.isArray((expression as OperationExpression).args)) {
    for (const arg of (expression as OperationExpression).args) {
      if ((arg as Term).termType === 'Variable') {
        for (const variable of Object.keys(variables)) {
          if ((arg as Term).value === variable) {
            const value = variables[variable];
            (arg as Term).termType = handleVariableType(value);
            (arg as Term).value = handleVariableValue(value);
          }
        }
      }
      if (Array.isArray((arg as OperationExpression).args)) {
        replaceExpression({ expression: arg as Expression, variables });
      }
    }
  }
  return expression;
};

const replaceTermWithVariables = ({
  term,
  variables,
}: {
  term: Term;
  variables: QueryVariables;
}): Term => {
  for (const variable of Object.keys(variables)) {
    replaceTerm({
      term,
      name: variable,
      value: variables[variable],
    });
    replaceTerm({
      term,
      name: variable,
      value: variables[variable],
    });
    replaceTerm({
      term,
      name: variable,
      value: variables[variable],
    });
  }
  return term;
};

const replaceWithVariables = ({
  triple,
  variables,
}: {
  triple: Triple;
  variables: QueryVariables;
}): Triple => {
  for (const variable of Object.keys(variables)) {
    replaceTerm({
      term: triple.subject,
      name: variable,
      value: variables[variable],
    });
    replaceTerm({
      term: triple.predicate as Term,
      name: variable,
      value: variables[variable],
    });
    replaceTerm({
      term: triple.object,
      name: variable,
      value: variables[variable],
    });
  }
  return triple;
};

const processPatterns = ({
  pattern,
  variables,
}: {
  pattern: Pattern | Quads;
  variables: QueryVariables;
}): Pattern | Quads => {
  // console.log(pattern);
  if (pattern.type === 'query') {
    sparqlWithSubstitution({
      query: pattern as Query,
      variables,
    });
  }
  if (pattern.type === 'graph') {
    replaceTermWithVariables({ term: pattern.name, variables });
  }
  if ((pattern as BindPattern).expression) {
    // console.log(pattern);
    // replace vars in anything that is not filter
    if (pattern.type !== 'filter') {
      replaceExpression({
        expression: (pattern as BindPattern).expression,
        variables,
      });
    }
  }
  // if pattern has triples - process directly
  if ((pattern as BgpPattern).triples) {
    for (const triple of (pattern as BgpPattern).triples) {
      replaceWithVariables({ triple, variables });
    }
  }
  // if pattern has patterns - process them recursively
  if ((pattern as GroupPattern).patterns) {
    for (const subPattern of (pattern as GroupPattern).patterns) {
      processPatterns({ pattern: subPattern, variables });
    }
  }
  return pattern;
};

export const sparqlWithSubstitution = ({
  query,
  variables = {},
}: {
  query: SparqlQuery;
  variables: QueryVariables;
}): SparqlQuery => {
  // traverse the tree and substitute variables if needed
  if (Object.keys(variables).length > 0) {
    // console.log(query);
    // process construct clause
    if ((query as ConstructQuery).template) {
      for (const pattern of (query as ConstructQuery).template) {
        replaceWithVariables({ triple: pattern, variables });
      }
    }

    // process where clause
    if ((query as Query).where) {
      for (const pattern of (query as Query).where) {
        processPatterns({ pattern, variables });
      }
    }

    // process update clause
    if ((query as Update).updates) {
      for (const updateEntry of (query as Update).updates) {
        // console.log({ updateEntry });
        if ((updateEntry as InsertDeleteOperation).graph !== undefined) {
          replaceTermWithVariables({
            term: ((updateEntry as InsertDeleteOperation).graph as any).name,
            variables,
          });
        }
        if ((updateEntry as InsertDeleteOperation).where) {
          for (const pattern of (updateEntry as InsertDeleteOperation).where) {
            processPatterns({ pattern, variables });
          }
        }
        if ((updateEntry as InsertDeleteOperation).delete) {
          for (const pattern of (updateEntry as InsertDeleteOperation).delete) {
            processPatterns({ pattern, variables });
          }
        }
        if ((updateEntry as InsertDeleteOperation).insert) {
          for (const pattern of (updateEntry as InsertDeleteOperation).insert) {
            processPatterns({ pattern, variables });
          }
        }
      }
    }
  }
  return query;
};
