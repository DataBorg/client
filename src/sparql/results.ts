import jsonld from 'jsonld';
import N3 from 'n3';
import type { SPARQLBindings, SPARQLResults } from 'types/client';

const prefixes = {
  'schema:': 'http://schema.org/',
  'rdf:': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
  'owl:': 'http://www.w3.org/2002/07/owl#',
  'dbo:': 'http://dbpedia.org/ontology/',
};

const replacePrefixes = (url: string) => {
  for (const prefix in prefixes) {
    const prefixUrl = prefixes[prefix];
    if (url.startsWith(prefixUrl)) {
      return url.replace(prefixUrl, prefix);
    }
  }

  return url;
};

const parseSimple = (results: SPARQLResults) => {
  const head = results.head?.vars;
  // if it's a boolean query with a boolean result
  // return the given boolean result as-is
  if (typeof results.boolean === 'boolean') {
    return results.boolean;
  }
  // otherwise - try to get bindings and process them
  const bindings = results.results?.bindings;
  const body = bindings?.map((binding) => {
    return head.reduce((acc, key) => {
      if (binding[key]) {
        acc[key] = replacePrefixes(binding[key].value);
      }
      return acc;
    }, {} as SPARQLResults);
  });
  return body;
};

const parseProperties = (
  results: SPARQLResults,
  predicates: SPARQLResultsPredicates
) => {
  const bindings = results.results.bindings;
  const resultMap = {};
  bindings.forEach((binding) => {
    const subject = replacePrefixes(binding[predicates.subjectVariable].value);
    const predicate = replacePrefixes(
      binding[predicates.predicateVariable].value
    );
    const object = replacePrefixes(binding[predicates.objectVariable].value);
    resultMap[subject] = resultMap[subject] || {};
    resultMap[subject][predicate] = object;
  });
  return resultMap;
};

const parseCustom = (
  results: SPARQLResults,
  rowProcessor: (row: SPARQLBindings, results: Array<any>) => any
) => {
  const bindings = results.results.bindings;
  const result = [];
  bindings.forEach((binding) => {
    rowProcessor(binding, result);
  });
  return result;
};

interface SPARQLResultsPredicates {
  subjectVariable?: string;
  predicateVariable?: string;
  objectVariable?: string;
}

interface SPARQLResultsParseOptions extends SPARQLResultsPredicates {
  type?: 'simple' | 'properties' | 'custom';
  rowProcessor?: (row: SPARQLBindings, results: Array<any>) => any;
}

export const parseSparqlResults = (
  results: SPARQLResults,
  options: SPARQLResultsParseOptions = {}
): any => {
  if (options.type === 'properties') {
    return parseProperties(results, options);
  }
  if (options.type === 'custom') {
    return parseCustom(results, options.rowProcessor);
  }
  // otherwise - fallback to simple processing
  return parseSimple(results);
};

const turtleToQuads = (turtle: string): Promise<any> => {
  return new Promise((resolve, reject) => {
    const parser = new N3.Parser();
    let parsed = [];
    parser.parse(turtle, (error, quad, prefixes) => {
      if (error) {
        reject(error);
        return;
      }
      if (quad) {
        parsed.push(quad);
        return;
      }
      resolve(parsed);
    });
  });
};

const quadsToString = (quads: Array<any>): Promise<string> => {
  return new Promise((resolve, reject) => {
    const writer = new N3.Writer({ format: 'N-Quads' });
    writer.addQuads(quads);
    writer.end((error, result) => {
      if (error) {
        reject(error);
        return;
      }
      resolve(result);
    });
  });
};

export const parseTextResults = async (
  results: string,
  prefixes: Record<string, string>,
  frame: object
): Promise<any> => {
  const quads = await turtleToQuads(results);
  const quadString = await quadsToString(quads);
  const doc = await jsonld.fromRDF(quadString, {
    format: 'application/n-quads',
  });
  // if user requested framing, apply it
  if (frame) {
    const framed = await jsonld.frame(doc, {
      '@context': prefixes ?? {},
      ...frame,
    });
    return framed;
  }
  return doc;
};
