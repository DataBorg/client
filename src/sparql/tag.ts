import { sparqlParser } from './parser';

function parseDocument(source: string) {
  const parsed = sparqlParser.parse(source);
  // console.log(parsed);
  if (!parsed) {
    throw new Error('Not a valid SPARQL query.');
  }
  return parsed;
}

function varPrefix() {
  return `PREFIX var: <var://>\n`;
}

export function sparql(
  literals: string | readonly string[],
  ...args: readonly any[]
) {
  // if input is one string - convert it to array
  // for further processing
  if (typeof literals === 'string') {
    literals = [literals];
  }

  // set first part of input as variable prefix
  // and initial input result
  let result = `${varPrefix()}${literals[0]}`;

  // parse all args and join them with rest of literals as a result
  args.forEach((arg, i) => {
    result += arg;
    result += literals[i + 1];
  });

  // console.log({ literals, args, result });
  return parseDocument(result);
}
