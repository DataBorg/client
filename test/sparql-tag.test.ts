import { describe, expect, it } from 'vitest';
import { sparql } from '../src/sparql/tag';

describe('sparql', () => {
  it('parses queries', () => {
    const query = sparql`SELECT ?s ?p ?o WHERE { ?s ?p ?o }`;
    expect(query).toMatchInlineSnapshot(`
      {
        "prefixes": {
          "var": "var://",
        },
        "queryType": "SELECT",
        "type": "query",
        "variables": [
          Variable {
            "termType": "Variable",
            "value": "s",
          },
          Variable {
            "termType": "Variable",
            "value": "p",
          },
          Variable {
            "termType": "Variable",
            "value": "o",
          },
        ],
        "where": [
          {
            "triples": [
              {
                "object": Variable {
                  "termType": "Variable",
                  "value": "o",
                },
                "predicate": Variable {
                  "termType": "Variable",
                  "value": "p",
                },
                "subject": Variable {
                  "termType": "Variable",
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
      }
    `);
  });

  it('throws an error on malformed query', () => {
    expect(() => {
      sparql`SELECT ?s ?p ?o WHERE ?s ?p ?o`;
    }).toThrowErrorMatchingSnapshot();
  });

  it('parses queries when called as a function', () => {
    expect(sparql('SELECT ?s WHERE { ?s ?p ?p }')).toMatchInlineSnapshot(`
      {
        "prefixes": {
          "var": "var://",
        },
        "queryType": "SELECT",
        "type": "query",
        "variables": [
          Variable {
            "termType": "Variable",
            "value": "s",
          },
        ],
        "where": [
          {
            "triples": [
              {
                "object": Variable {
                  "termType": "Variable",
                  "value": "p",
                },
                "predicate": Variable {
                  "termType": "Variable",
                  "value": "p",
                },
                "subject": Variable {
                  "termType": "Variable",
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
      }
    `);
  });

  it('parses queries with weird substitutions', () => {
    const obj = Object.create(null);
    expect(sparql`SELECT ?s WHERE { ?s ?p "${obj.missing}" }`)
      .toMatchInlineSnapshot(`
        {
          "prefixes": {
            "var": "var://",
          },
          "queryType": "SELECT",
          "type": "query",
          "variables": [
            Variable {
              "termType": "Variable",
              "value": "s",
            },
          ],
          "where": [
            {
              "triples": [
                {
                  "object": Literal {
                    "datatype": NamedNode {
                      "termType": "NamedNode",
                      "value": "http://www.w3.org/2001/XMLSchema#string",
                    },
                    "language": "",
                    "termType": "Literal",
                    "value": "undefined",
                  },
                  "predicate": Variable {
                    "termType": "Variable",
                    "value": "p",
                  },
                  "subject": Variable {
                    "termType": "Variable",
                    "value": "s",
                  },
                },
              ],
              "type": "bgp",
            },
          ],
        }
      `);
    expect(sparql`SELECT ?s WHERE { ?s ?p "${null}" }`).toMatchInlineSnapshot(`
      {
        "prefixes": {
          "var": "var://",
        },
        "queryType": "SELECT",
        "type": "query",
        "variables": [
          Variable {
            "termType": "Variable",
            "value": "s",
          },
        ],
        "where": [
          {
            "triples": [
              {
                "object": Literal {
                  "datatype": NamedNode {
                    "termType": "NamedNode",
                    "value": "http://www.w3.org/2001/XMLSchema#string",
                  },
                  "language": "",
                  "termType": "Literal",
                  "value": "null",
                },
                "predicate": Variable {
                  "termType": "Variable",
                  "value": "p",
                },
                "subject": Variable {
                  "termType": "Variable",
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
      }
    `);
    expect(sparql`SELECT ?s WHERE { ?s ?p "${0}" }`).toMatchInlineSnapshot(`
      {
        "prefixes": {
          "var": "var://",
        },
        "queryType": "SELECT",
        "type": "query",
        "variables": [
          Variable {
            "termType": "Variable",
            "value": "s",
          },
        ],
        "where": [
          {
            "triples": [
              {
                "object": Literal {
                  "datatype": NamedNode {
                    "termType": "NamedNode",
                    "value": "http://www.w3.org/2001/XMLSchema#string",
                  },
                  "language": "",
                  "termType": "Literal",
                  "value": "0",
                },
                "predicate": Variable {
                  "termType": "Variable",
                  "value": "p",
                },
                "subject": Variable {
                  "termType": "Variable",
                  "value": "s",
                },
              },
            ],
            "type": "bgp",
          },
        ],
      }
    `);
  });

  it('returns the same object for the same query, even with whitespace differences', () => {
    expect(sparql`SELECT ?s WHERE { ?s ?p ?o }`).toEqual(
      sparql`SELECT ?s WHERE { ?s ?p ?o }`
    );
  });
});
