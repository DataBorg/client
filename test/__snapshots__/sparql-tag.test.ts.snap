// Vitest Snapshot v1

exports[`sparql > throws an error on malformed query 1`] = `
"Parse error on line 2:
...LECT ?s ?p ?o WHERE ?s ?p ?o
-----------------------^
Expecting '{', got 'VAR'"
`;
